bench/benches/operators.rs

```diff
@@ -0,0 +1,178 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+use nb2::batch::Either;
+use nb2::packets::{Ethernet, Packet};
+use nb2::testils::criterion::BencherExt;
+use nb2::testils::proptest::*;
+use nb2::{compose, Batch, Mbuf};
+use proptest::prelude::*;
+use proptest::strategy;
+
+const BATCH_SIZE: usize = 500;
+static mut COUNTER: u32 = 0;
+
+fn filter_true(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    batch.filter(|_p| true)
+}
+
+fn filter_false(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    batch.filter(|_p| false)
+}
+
+fn filter_map(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Ethernet> {
+    batch.filter_map(|p| {
+        let ethernet = p.parse::<Ethernet>()?;
+        Ok(Either::Keep(ethernet))
+    })
+}
+
+fn map_then_filter(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Ethernet> {
+    batch.map(|p| p.parse::<Ethernet>()).filter(|_p| true)
+}
+
+fn map(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Ethernet> {
+    batch.map(|p| p.parse::<Ethernet>())
+}
+
+fn group_by(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    unsafe { COUNTER += 1 };
+
+    unsafe {
+        batch.group_by(
+            |_p| COUNTER,
+            |groups| {
+                compose!(groups {
+                    0 => |group| {
+                        group.for_each(|_p| Ok(()))
+                    }
+                    _ => |group| {
+                        group.for_each(|_p| Ok(()))
+                    }
+                })
+            },
+        )
+    }
+}
+
+fn replace(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    batch.replace(|_p| Mbuf::new())
+}
+
+fn no_batch_replace(_mbuf: Mbuf) -> Mbuf {
+    Mbuf::new().unwrap()
+}
+
+#[nb2::bench(mempool_capacity = 511)]
+fn filters_batch(c: &mut Criterion) {
+    let mut group = c.benchmark_group("operators::filter");
+
+    group.bench_function("operators::filter_true", |b| {
+        let s = v4_udp();
+        b.iter_proptest_polled(s, filter_true, BATCH_SIZE)
+    });
+
+    group.bench_function("operators::filter_false", |b| {
+        let s = v4_udp();
+        b.iter_proptest_polled(s, filter_false, BATCH_SIZE)
+    });
+
+    group.finish()
+}
+
+#[nb2::bench(mempool_capacity = 511)]
+fn filter_map_vs_map_then_filter_batch(c: &mut Criterion) {
+    let mut group = c.benchmark_group("operators::filter_map_vs_map_then_filter");
+
+    group.bench_function("operators::filter_map", |b| {
+        let s = strategy::Union::new_weighted(vec![(9, v4_udp().boxed()), (1, v6_udp().boxed())]);
```

Comment:
don't think this union here add any meaningful variance to the result since we do go down far enough to make v4/v6 matter when benching the combinators, by design of course.

Reply:
It hits the error path is all.